contracts/TokenCollector.sol

```diff
@@ -0,0 +1,99 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.7.6;
+pragma abicoder v2;
+
+import { IERC20Permit } from "@openzeppelin/contracts/drafts/IERC20Permit.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+
+import { ISpender } from "contracts/interfaces/ISpender.sol";
+import { ITokenCollector } from "contracts/interfaces/ITokenCollector.sol";
+import { IUniswapPermit2 } from "contracts/interfaces/IUniswapPermit2.sol";
+import { SpenderLibEIP712 } from "contracts/utils/SpenderLibEIP712.sol";
+
+contract TokenCollector is ITokenCollector {
+    using SafeERC20 for IERC20;
+
+    address public immutable owner;
+    address public immutable spender;
+    address public immutable uniswapPermit2;
+
+    constructor(address _spender, address _uniswapPermit2) {
+        owner = msg.sender;
+        spender = _spender;
+        uniswapPermit2 = _uniswapPermit2;
+    }
+
+    function collect(
+        address token,
+        address from,
+        address to,
+        uint256 amount,
+        bytes memory data
+    ) external override {
+        require(msg.sender == owner, "TokenCollector: not owner");
```

Comment:
If the owner (generic swap?) allows arbitrary message call, then it need to forbid arbitrary calls to this contract.

Reply:
Yes, for strategy supports arbitrary call it should responsible to block others to access token collector. Just like CowSwap does.

https://github.com/cowprotocol/contracts/blob/a7cb2900191c4871b72f2a8f3f81ba4029d14070/src/contracts/GPv2Settlement.sol#L456-L459